Lesson 5 HW

grrrvahrrr opened the pull request:
Задание 5. В закомментированном коде способ вычисление числа фибоначчи с помощью рекурсии, а потом с помощью мап. Не знаю, правильно ли я понял задание, но совместить два способа у меня не получилось, хотя мучался долго) с интернетом не советовался в этот раз, только с документацией
Мне кажется слайс здесь было бы удобнее использовать нежели чем мап, так как слайс уже и так пронумерован, а в данном случае "имена" цифрам не придумать.

grrrvahrrr committed:
Lesson 5 HW

goarchitecture commented on lesson5/main.go:
```diff
@@ -0,0 +1,84 @@
+	"os"
+)
+
+func fibb(x uint) uint {
```
не надо комментировать код. Вы могли создать 2 функции, просто назвав их по-разному. Одна бы считала рекурсией, вторая - рекурсией с мапой

goarchitecture commented on lesson5/main.go:
```diff
@@ -0,0 +1,84 @@
+)
+
+func fibb(x uint) uint {
+	if x <= 0 {
```
ваш тип `uint`, проверка `if x <= 0` вводит в заблуждение. на самом деле вы проверяете `if x == 0`

goarchitecture commented on lesson5/main.go:
```diff
@@ -0,0 +1,84 @@
+		os.Exit(1)
+	}
+
+	if num < 0 {
```
не возможно, ошибка выйдет ещё на скане (так как отрицательно число нельзя записать в uint)

goarchitecture commented on lesson5/main.go:
```diff
@@ -0,0 +1,84 @@
+func fibbMapCalc(x int) int {
+	fibbMap := make(map[int]int, x+1)
+
+	fibbMap[0] = 0
```
вы писали в комментарии что лучше слайс чем мапа:
со слайсом (точнее именно в этом случае с массивом ) - можно, но легко ошибится и выпасть в панику. Конкретно здесь, легче с мапой работать

goarchitecture commented on lesson5/main.go:
```diff
@@ -0,0 +1,84 @@
+	fibbMap[0] = 0
+	fibbMap[1] = 1
+
+	for i := 2; i < x+1; i++ {
```
если уходите от рекурсии то можно и вовсе без мапы. хранить значения prev1, prev2 и всё.
Попробуйте всё-таки добавить мапу в версию с рекурсией

goarchitecture reviewed (changes requested):
Хорошо. Доработайте по комментариям.

grrrvahrrr committed:
Lesson 5 HW Update

Меня кажется осенило) Разложил все поэтапно, добавил аргументы и получилась функция fibbMapRecursion, которая рекурсивно считает с помощью мап.

goarchitecture commented on lesson5/main.go:
```diff
@@ -64,6 +36,24 @@ func fibbMapCalc(x int) int {
 
+func fibbMapRecursion(numMap map[int]int, x int) (map[int]int, int) {
+
+	numMap[0] = 0
```
вы каждый раз при вызове этой функции будете устанавливать одно и тоже?

goarchitecture commented on lesson5/main.go:
```diff
@@ -64,6 +36,24 @@ func fibbMapCalc(x int) int {
+	numMap[1] = 1
+	length := len(numMap)
+
+	numMap[length] = numMap[length-1] + numMap[length-2]
```
к примеру я вызываю fibbMapRecursion({}, 10)
функция заполняет 0, 1 (выше), затем заполняет 2 (так как length = 2)
затем функция просто вызовет себя ещё раз - и тогда несколько раз пока не дойдет до 10... 

не рациональная система

grrrvahrrr replied:
Добавил еще один коммит, немного переписал функцию. Но я, наверное, не очень понимаю задание. Нам нужно написать рекурсивную функцию с мапой, то есть она должна работать по кругу пока не заполнит мапу. В новом коммите я вынес задание первых трех значений в мейн, а нулевое значение стало необходимой нам длинной. Сама рекурсия доходит до заданной длины и останавливается. Каждый раз совершая определение следующего значения в мапе, функция просто берет два предыдущих значения из мапы.
Мне специально не хочется писать функцию, которая каждый раз создавала бы новую мапу, так как это не оптимально, поэтому и выношу ее создание за пределы функции, а в саму функцию принимаю мапу.

grrrvahrrr replied:
В последней (10ой) лекции подсмотрел решение этой задачки у вас в коде во время вебинара) сложно) честно говоря я даже не думал в этом направлении и решение искал в других местах. Теперь раз уж так интересно написать тесты на все свои варианты и посмотреть, что у меня получилось в плане оптимизации) сделаю это в 10ой домашке

goarchitecture reviewed (commented):
"мудренно". попробуйте разделить получше кеш и рекурсивную функцию

grrrvahrrr committed:
Lesson 5 HW Update 2

grrrvahrrr committed:
Bug hunting

goarchitecture reviewed (approved):
В общем - принято.
На самом деле всё проще чем кажется. Да, я на одной из последних лекций показал пример, как "добавить мапу" в эту функцию.
Ведь идея практическая (и то, что реально применимо и может быть подобное использоваться в реальном мире) - то, что сначала у вас простая рекурсивная функция.

Но она оказалась медленная. Если вы вызываете fib(7) то она ведет так, скажет, 2 крыла: fib(6) и fib(5). А каждое из этих 2 вызовов открывает свои 2 крылам. И т.д.
Получается такое бинарное дерево вызовов. Каждый раз количество вывзовов функции получается, что умножается на 2 fib(40) будет вызывать самого себя 2^40 раз. Это огромное число. И поэтому получается медленно.

Если на эту рекурсию взнлянуть внимательно, то можно увидеть, что большинство вызовов - повторяются. Это можно исправить мапой.
Она будет является так называемым кешем. И да, она может быть как глобальная - то есть чтобы между разными вызовами функции fib() - использовался один и тот же  кеш - так и локальная (заинкапсулированная в структуру, либо в замыкание).
иде я в том, что раз получив какое-то число фтбоначии - мы просто сохраняем его, и больше не вычисляем его снова.
